beartype/_protocol.py

```diff
@@ -0,0 +1,277 @@
+
+        .. code-block:: python
+          :linenos:
+
+          >>> from abc import abstractmethod
+          >>> from beartype import beartype
+          >>> from beartype.typing import Protocol
+
+          >>> class MyBearProtocol(Protocol):  # <-- runtime-checkable through inheritance
+          ...   @abstractmethod
+          ...   def myfunc(self, arg: int) -> str:
+          ...     pass
+
+          >>> my_thing: MyBearProtocol = MyImplementation()
+          >>> isinstance(my_thing, MyBearProtocol)
+          True
+
+          >>> @beartype
+          ... def do_somthing(thing: MyBearProtocol) -> None:
+          ...   thing.myfunc(0)
+        """
+
+        __slots__: Union[str, Iterable[str]] = ()
+
+        def __class_getitem__(cls, params):
+            # We have to redefine this method because typing.Protocol's
+            # version is very persnickety about only working for
+            # typing.Generic and typing.Protocol. That's an exclusive club,
+            # and we ain't in it. (RIP, GC.)
+
+            # We *can* however, call typing.Protocol's implementation directly
+            # to get the resulting generic alias.
+            fake_gen_alias = _Protocol.__class_getitem__(params)
+
+            # It has two teensie weensie little problems at this point,
+            # though. First, it's __origin__ attribute is wrong for our
+            # purposes. It's (unsurprisingly) typing.Protocol, but we need it
+            # to be our class. Otherwise our inheritors end up with the wrong
+            # metaclass (i.e., type(typing.Protocol) instead of the desired
+            # _CachingProtocolMeta), for some reason. (Luddite alert: I don't
+            # fully understand the mechanics here. I suspect no one really
+            # does.) Second, typing.Protocol.__class_getitem__ is memoized
+            # with an LRU cache. This means doing surgery on the returned
+            # value could show up for others, including those not even using
+            # our Protocol implementation! *SIGH* What's a bear to do?
+
+            # The good news is that the generic alias provides us a way of
+            # making a kind of copy. (Yay!)
+            new_gen_alias = fake_gen_alias.copy_with(params)
+
+            # The bad news is that it's incomplete. copy_with does *not* copy
+            # _typevar_types and _paramspec_tvars, which were introduced in
+            # Python 3.10. (Boo!)
+            if hasattr(fake_gen_alias, "_typevar_types"):
+                new_gen_alias._typevar_types = fake_gen_alias._typevar_types
+            if hasattr(fake_gen_alias, "_paramspec_tvars"):
+                new_gen_alias._paramspec_tvars = fake_gen_alias._paramspec_tvars
+
+            # Now all we have to do is replace the __origin__ in our clone to
+            # point to ourself.
+            new_gen_alias.__origin__ = cls
+
+            # We're done! Time for a honey brewskie break. We earned it.
+            return new_gen_alias
```

Comment:
One last attempt….

Reply:
Hmmm…maybe not? `_typevar_types` might influence how `params` is interpreted on construction. That `typing.py` is full of some obfuscated stuff! Who can maintain anything like that?

Anyway, I think what we want is something like this:

``` python
if hasattr(fake_gen_alias, "_typevar_types"):
  new_gen_alias = type(fake_gen_alias)(cls, params,
    name=fake_gen_alias._name, inst=fake_gen_alias._inst,
    _typevar_types=fake_gen_alias._typevar_types,
    _paramspec_tvars=fake_gen_alias._paramspec_tvars)
else:
  new_gen_alias = type(fake_gen_alias)(cls, params,
    name=fake_gen_alias._name, inst=fake_gen_alias._inst)
```

Reply:
To bypass the LRU cache, you could fetch `__wrapped__` on the func, which `@wraps` sets for you. Would that be easier?

Reply:
Yeah, I thought about that. I initially rejected it in case `@_tp_cache` ever went away, but I think that fear is pretty silly. I'm already up to my eyeballs in seven-layer-bean-and-private-implementation-details dip. Let me give it a shot. I think it will be cleaner.